### PR TITLE
feat(workflows): package workflow resources

### LINKED ISSUE #2527: Goal: workflows are self-contained, copyable folders
## Goal

A workflow can be authored, shared, pinned, and copied as one self-contained folder containing its definition, commands, and scripts.

Authors choose both directory names:

```text
.archon/workflows/<pack>/<workflow>/
  <workflow definition>.yaml
  commands/
  scripts/
```

The same structure works for bundled workflows, home-scoped workflows, and workflows committed to a target repository. Neither `defaults` nor any other pack name is reserved.

## Why this matters

The current filesystem layout splits one logical workflow across `.archon/workflows/`, `.archon/commands/`, and `.archon/scripts/`. The video workflow set demonstrates the cost: five related YAML files live together while their six scripts are stored in a shared directory.

Repository copy is Archon's pinning/distribution mechanism. Copying a workflow should mean copying one folder, not finding related files across several directories.

This decision is upstream of the planned bundled-workflow rewrite in #2123. Authoring that rewrite into the flat layout first would create a second migration immediately afterward.

## User outcome

A workflow author can:

1. Pick any pack name and workflow folder name.
2. Keep the workflow YAML and its private commands/scripts together.
3. Copy that folder into another repository or home scope.
4. Run it without rewriting local resource references or moving files into shared directories.
5. Compose workflows while included workflows continue to resolve their own resources.

## Definition of done

- [ ] Repository workflows support `.archon/workflows/<pack>/<workflow>/`.
- [ ] Home-scoped workflows support the identical structure.
- [ ] Bundled workflows are authored and generated from the identical structure.
- [ ] Pack and workflow directory names are author-defined; `defaults` has no special package semantics.
- [ ] Commands and named scripts resolve relative to their owning workflow folder.
- [ ] Included workflow nodes retain the included workflow's resource ownership.
- [ ] Different workflows can reuse the same command and script basenames without collisions.
- [ ] Existing workflow `name:` uniqueness and node execution semantics remain unchanged.
- [ ] Existing repo-over-home-over-bundled precedence remains deterministic.
- [ ] Same-scope ambiguous workflow filename collisions fail with an actionable error rather than depending on filesystem order.
- [ ] Generated bundles include packaged YAML, commands, and executable named scripts.
- [ ] Existing flat workflow/resource layouts have an explicit compatibility or migration path.
- [ ] CLI validation, listing, execution, source mode, and compiled-binary mode have coverage.
- [ ] User-facing workflow authoring and copy/install documentation reflects the packaged layout.

## Constraints / non-goals

- Do not require author-visible qualified command or script names.
- Do not add a workflow-language field solely to carry filesystem provenance.
- Do not change scheduling, substitution, output, timeout, dependency, or failure semantics.
- Do not use unbounded recursive workflow discovery; the two directory levels are the package boundary.
- Do not make `defaults` structurally special.

## Spike evidence

A falsification spike reached a **CONDITIONAL** verdict: the goal works without public YAML or engine-semantic changes, provided discovery and bundle generation understand the fixed two-level package boundary and resource ownership is carried internally.

Evidence and reproduction artifacts:

https://gist.github.com/Wirasm/0013edcb72bba1e7bd043f231045270f

The spike exercised:

- Actual Archon repo and home discovery.
- Direct and included command/script resolution with colliding basenames.
- Included-block resource ownership.
- Exact-filename repo override behavior.
- Bundle indexing and bundled-script materialization/execution.
- The real five-workflow/six-script video fixture.
- Repeated runs and `@archon/workflows` type-check.

The current generic loader and bundle generator cannot see the requested two-level arbitrary-pack tree unchanged; that is the named condition to address.

### PR DESCRIPTION
**Pull request by Wirasm**
## Summary

- Problem: workflow YAML, commands, and scripts are split across shared directories, so copying one workflow requires a scavenger hunt.
- Why it matters: repo-copy is Archon's pinning/distribution mechanism, and the upcoming bundled workflow rewrite should author directly into the final package layout.
- What changed: discovers `.archon/workflows/<pack>/<workflow>/`, binds bare local command/script references to their exact owner, embeds packaged resources in binaries, preserves packaged GET/PUT/DELETE lifecycle in the Web API, and documents the layout for repository and home scopes.
- What did **not** change (scope boundary): authored YAML syntax, workflow engine node semantics, executor signatures, cross-scope filename override rules, and legacy flat layouts remain unchanged.

### Before

```
Author                         Archon
──────                         ──────
creates workflow YAML ──────▶  .archon/workflows/
creates commands ───────────▶  .archon/commands/
creates scripts ────────────▶  .archon/scripts/
copies a workflow pack ─────▶  gathers files from three trees
```

### After

```
Author                         Archon
──────                         ──────
creates one folder ─────────▶  [.archon/workflows/<pack>/<workflow>/]
  workflow YAML                [discovers fixed package boundary]
  commands/                    [resolves against exact workflow owner]
  scripts/                     [resolves/materializes exact owner]
copies one folder ──────────▶  [self-contained workflow arrives intact]
```

### Before

```
workflow-discovery ──▶ loader ──▶ include-expander
executor-shared ─────▶ shared command directories
script-discovery ────▶ shared script directories
bundle generator ────▶ workflows + commands generated records
validator ───────────▶ shared command/script discovery
```

### After

```
workflow-discovery [~] ===▶ packaged-workflow [+] ===▶ include-expander
          │                         │
          │                         ├===▶ executor-shared [~] ──▶ exact commands/
          │                         ├===▶ script-discovery [~] ──▶ exact scripts/
          │                         └===▶ validator [~]
          │
bundle generator [~] ===▶ workflows + owners + commands + scripts records [~]
                                  │
                                  └===▶ bundled script materialization [+]
```

**Connection inventory** (list every module-to-module edge, mark changes):

| From | To | Status | Notes |
|------|----|--------|-------|
| workflow discovery | packaged resource module | **new** | Qualifies bare local references at load time |
| workflow discovery | include expander | **modified** | Included nodes retain the child package owner |
| executor shared | packaged resource module | **new** | Resolves commands from the exact source/pack/workflow |
| script discovery | packaged resource module | **new** | Keys scripts by exact owner |
| script discovery | bundled generated data | **modified** | Materializes embedded scripts to deterministic cache paths |
| validator | packaged resource module | **new** | Validates local command/script presence through the same identity |
| bundle generator | bundled generated data | **modified** | Emits owner, command, and script records deterministically |

## Label Snapshot

- Risk: `risk: high`
- Size: `size: XL`
- Scope: `workflows|server|docs|tests`
- Module: `workflows:discovery`, `workflows:executor`, `workflows:validator`, `workflows:defaults`

## Change Metadata

- Change type: `feature`
- Primary scope: `workflows`

## Linked Issue

- Fixes #2527
- Related #2123
- Depends on: none
- Supersedes: none

## Validation Evidence (required)

Commands and result summary:

```bash
bun install --frozen-lockfile                         # passed
bun run check:bundled                                # passed: 37 commands, 21 workflows, 0 packaged scripts in-tree
bun run check:bundled-skill                          # passed
bun run check:bundled-schema                         # passed
bun run check:pi-vendor-map                          # passed
bun run check:capability-matrix                      # passed
bun run type-check                                   # passed, all packages
bun x eslint packages/workflows/src scripts/generate-bundled-defaults.ts --max-warnings 0
                                                        # passed
bun run format:check                                 # passed
bun run test:install                                 # passed
bun run test                                         # passed, all packages and scripts
bun run validate                                     # reaches lint, then fails on 2 pre-existing provider lint errors
```

- Evidence provided: real-filesystem discovery/resolution tests, include-owner tests, collision tests, generated-record completeness tests, validator tests, and a compiled-bundle script materialization test that executes the resulting file.
- `bun run validate` is not fully green because strict lint reports `no-unsafe-call` in `packages/providers/src/community/pi/provider.ts:163` and `packages/providers/src/shared/structured-output.ts:112`; both files are byte-identical to `origin/dev`. The changed surface passes strict lint, and every subsequent gate was run separately and passed.

## Security Impact (required)

- New permissions/capabilities? `No`
- New external network calls? `No`
- Secrets/tokens handling changed? `No`
- File system access scope changed? `Yes`
- Packaged discovery reads exactly two author-controlled directory levels under existing workflow roots. Generated scripts materialize under Archon's cache using validated path segments, a content hash, temporary write, and atomic rename. Traversal and separator-bearing names are rejected.

## Compatibility / Migration

- Backward compatible? `Yes`
- Config/env changes? `No`
- Database migration needed? `No`
- Existing flat workflow, command, and script paths continue to work. Authors can opt into the package layout one workflow at a time by moving its YAML and local resources into one folder.

## Human Verification (required)

- Verified scenarios: arbitrary author pack/workflow names in repo and home scopes; exact-owner commands/scripts; include ownership; loop and loop-group qualification; repo override ownership; bundled generation and executable script materialization; packaged Web fetch/update/delete.
- Edge cases checked: local basename collisions, same-scope YAML filename collisions, invalid path segments, missing local resource without cross-scope fallback, inline scripts remaining inline, nested bundled scripts, untracked bundled inputs, deterministic repeated materialization, source-bundled immutability, complete package deletion, and filename/declared-name collisions.
- What was not verified: no packaged workflow has been added to the shipping `.archon/workflows/` tree in this PR; generator integration is covered with temporary fixtures instead.

## Side Effects / Blast Radius (required)

- Affected subsystems/workflows: workflow discovery, include expansion inputs, command prompt loading, named script discovery, resource validation, bundled-default generation.
- Potential unintended effects: internal qualified resource references could surface through consumers that inspect post-discovery workflow definitions; malformed package-shaped directories now produce explicit discovery errors.
- Guardrails/monitoring for early detection: fixed-depth scanning, exact-owner resolution, collision errors, generated completeness checks, and full legacy test coverage.

## Rollback Plan (required)

- Fast rollback command/path: revert commit `cde7ff8f`; legacy flat layouts remain intact and require no data migration.
- Feature flags or config toggles: none.
- Observable failure symptoms: packaged workflows omitted from listing, local command/script validation errors, or bundled script materialization errors with owner-qualified context.

## Risks and Mitigations

- Risk: identical local resource names in different packs could resolve to the wrong file.
  - Mitigation: source, pack, and workflow are encoded in an internal validated identity and resolution never falls through to another owner.
- Risk: binary scripts need a real path without changing executor semantics.
  - Mitigation: embedded content is materialized to a content-hashed cache path with atomic rename, preserving the existing `ScriptDefinition.path` contract.
- Risk: recursive discovery could accidentally accept arbitrary layouts.
  - Mitigation: the package scanner accepts exactly `<pack>/<workflow>/` and exactly one YAML/YML file per workflow folder.

## Summary by CodeRabbit

* **New Features**
  * Added packaged workflows in customizable pack/workflow directories.
  * Packaged workflows can include YAML definitions, commands, scripts, and ownership information.
  * Added resource resolution across bundled, repository, and home-scoped workflows.
  * Repository workflows override matching bundled defaults.
  * Bundled scripts support stable cached execution paths.
  * Workflow creation, updates, and deletion support packaged layouts.

* **Compatibility**
  * Flat and one-level grouped layouts remain supported.
  * Deeper nesting is ignored.
  * Existing command and script references continue to work.

* **Documentation**
  * Updated workflow, command, script, and directory layout guidance.

### CODE REVIEW

**Comment by Wirasm** [comment]
## PR Review Summary

### Critical Issues (0 found)

None. The code reviewer’s initial blockers were fixed and re-reviewed:

- Bundled generation now embeds one-level nested packaged scripts, matching source discovery.
- Packaged workflows are fetchable and mutable through the Web API without flattening.
- Source-bundled packages remain immutable while flat source-checkout workflows stay project-owned.
- Filename fast paths verify declared workflow names in filesystem and embedded records.

### Important Issues (0 found)

None. Packaged deletion now removes the complete workflow folder, and legacy grouped workflows accept both basename and full namespaced declarations.

### Suggestions (0 found)

All five simplification suggestions were applied: explicit source/root and error branches, explicit runtime detection, simpler collision merging, and centralized idempotent resource qualification.

### Strengths

- Exact ownership includes source, pack, and workflow, preventing cross-package basename collisions and silent cross-scope fallback.
- Qualification covers commands, loop commands, named scripts, loop-group children, and included workflow ownership.
- Binary scripts use content-addressed paths with atomic materialization.
- Source and compiled bundle behavior now match for direct and one-level nested scripts.
- Web GET/PUT/DELETE preserves raw authored YAML, package location, source ownership, and complete package lifecycle.
- Regression coverage includes repo/home packages, overrides, collisions, includes, validation, bundle generation, executable materialization, Web lifecycle, bundled immutability, and filename/declared-name mismatches.

### Documentation Issues

None. README, CLAUDE.md, workflow authoring docs, and script-node docs describe the package layout and legacy compatibility.

### Verdict

READY TO MERGE

### Recommended Actions

1. Let the refreshed GitHub Actions checks complete on commit `28f6009d`.
2. Merge after required checks are green.

**Comment by Wirasm** [comment]
## PR Review Summary

### Critical Issues (1 found)

| Agent | Issue | Location |
|---|---|---|
| errors | Bundle generation silently treats every `stat()` failure as “not a directory,” so EACCES/EIO can ship an incomplete embedded bundle while validation reports success. Only ENOENT may be treated as absence. | `scripts/generate-bundled-defaults.ts:237` |

### Important Issues (8 found)

| Agent | Issue | Location |
|---|---|---|
| code | Windows paths from `relative()` use backslashes while Git reports `/`; tracked packaged files are misclassified as untracked. Disable Git path quoting too. This is the current Windows CI failure. | `scripts/generate-bundled-defaults.ts:175` |
| errors | Packaged script discovery suppresses permission/I/O failures, later reporting scripts as merely missing. | `packages/workflows/src/script-discovery.ts:169` |
| errors | Packaged workflow discovery silently drops inaccessible pack/workflow entries. | `packages/workflows/src/workflow-discovery.ts:226`, `:254` |
| errors | Include safety scanning converts every packaged-command read error to absence, which can bypass sibling-reference scanning. | `packages/workflows/src/workflow-discovery.ts:404` |
| code | Packaged command validation accepts a directory named `<command>.md`; runtime then fails with EISDIR. | `packages/workflows/src/validator.ts:217` |
| code | Per-entry packaged API lookup does not tolerate ENOENT races, while malformed target YAML is silently converted to 404 and PUT may create a second flat workflow. | `packages/server/src/routes/api.ts:149` |
| errors | Bundled duplicate-name detection can throw outside DELETE's structured error boundary. | `packages/server/src/routes/api.ts:4037` |
| docs | User docs still describe commands/scripts as shared fallback chains and workflow discovery as recursive; packaged resources are owner-local with no fallback and the package layout has exact depth. | `packages/docs-web/src/content/docs/guides/authoring-workflows.md`, `script-nodes.md`, `global-workflows.md` |

### Suggestions (7 found)

| Agent | Issue | Location |
|---|---|---|
| tests | Add mutation-safety tests for duplicate declared workflow names and packaged API failure behavior. | `packages/server/src/routes/api.workflows.test.ts` |
| tests | Add ENOENT/EACCES tests for packaged script discovery and a flat-vs-packaged collision test. | `packages/workflows/src/script-discovery.test.ts`, `loader.test.ts` |
| tests | Add binary empty-command parity and directory-as-command validation coverage. | `load-command-prompt.test.ts`, `validator.test.ts` |
| tests | Give subprocess-heavy generator tests an explicit Windows-safe timeout. | `packages/workflows/src/defaults/generate-bundled-defaults.test.ts` |
| comments | Update stale module docs describing only shared script/default sources. | `packages/workflows/src/script-discovery.ts`, `packages/workflows/src/defaults/bundled-defaults.ts`, `scripts/generate-bundled-defaults.ts` |
| types | Express bundled script extension/runtime as a discriminated union and ownership metadata as a sparse keyed record. | `scripts/generate-bundled-defaults.ts` |
| types | Validate packaged-reference construction so the formatter cannot emit references its parser rejects. | `packages/workflows/src/packaged-workflow.ts` |

### Strengths

- The ownership qualification model is centralized and covers nested loop-group resources.
- Project, home, and bundled package resources are isolated rather than falling through across owners.
- The generator emits workflow ownership, commands, scripts, and deterministic content.
- Existing tests cover packaged GET/PUT/DELETE happy paths, bundled protection, binary script materialization, and stable cache reuse.
- The Windows matrix correctly caught a real portability defect.

### Documentation Issues

- Clarify the exact `<pack>/<workflow>/<yaml>` layout and invalid deeper nesting.
- Split command/script resolution docs into packaged owner-local mode versus legacy shared precedence.
- Replace claims of arbitrary recursive workflow discovery with the supported flat, grouped, and packaged layouts.

### Verdict

**NEEDS FIXES** — the Windows failure and fail-fast/resource-resolution issues block merge.

### Recommended Actions

1. Fix path normalization and unexpected filesystem-error handling.
2. Make validation and API lookup preserve file/type/error invariants.
3. Add focused regression tests for the corrected contracts and Windows timeout.
4. Update the canonical user docs and stale module comments.
5. Re-run targeted specialist reviews and `bun run validate`.

**Comment by Wirasm** [comment]
## PR Review Summary

### Critical Issues (0 found)

All critical findings were fixed and passed targeted re-review.

### Important Issues (0 found)

All important findings were fixed and passed code, error-handling, and documentation re-review.

### Suggestions (2 deferred)

| Agent | Issue | Disposition |
|---|---|---|
| tests | Add separate generator tests for every duplicate guard. | Deferred: runtime collision tests and generator integration coverage now protect the changed behavior; the remaining duplicate branches are existing fail-fast guards. |
| review | Exclude a pack named `defaults`. | Rejected: `defaults` is intentionally not reserved for authored packages; only the explicit legacy compatibility path is special. |

### Strengths

- Windows-safe Git path comparison now normalizes separators and disables quoted path output.
- Packaged workflow, command, and script discovery fails fast on permission/I/O errors while tolerating ENOENT races.
- GET/PUT/DELETE preserve malformed-package and duplicate-name safety; mutation tests verify no package is changed on ambiguity.
- Packaged command validation requires real files, binary commands preserve empty-file semantics, and runtime failures are contextually logged.
- Generated script metadata encodes valid extension/runtime pairs, and ownership metadata is typed as sparse.
- Documentation consistently distinguishes owner-local packaged resources from legacy shared precedence.

### Documentation Issues

None remaining after re-review.

### Verdict

**APPROVE** — no critical or important findings remain.

### Validation

- `bun run validate` — passed in full.
- Targeted re-reviews: code PASS, error-handling PASS, documentation PASS.
- Focused regression suites passed, including API workflow lifecycle, generator, validator, discovery, loader, and bundled command tests.
- Fix commit: `4cdd0743c8ae6416e416f2a565e756aebacc8c6e`.

### Recommended Actions

1. Let the refreshed GitHub Actions matrix complete.
2. Merge once required checks are green.

File: packages/workflows/src/workflow-discovery.ts [hunk 1]
```diff
@@ -174,6 +205,127 @@ async function loadWorkflowsFromDir(dirPath: string, depth = 0): Promise<DirLoad
...
+  }
+
+  for (const pack of packs.sort((a, b) => a.localeCompare(b))) {
+    const packPath = join(workflowsRoot, pack);
+    try {
+      if (!(await stat(packPath)).isDirectory()) continue;
+    } catch {
+      continue;
+    }
+    if (!isValidWorkflowFolderSegment(pack)) {
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply packages/workflows/src/workflow-discovery.ts:246]: Keeping this behavior intentionally. Issue #2527 explicitly requires that neither `defaults` nor any other pack name is reserved, and its non-goals say not to make `defaults` structurally special. Legacy defaults discovery reads direct YAML files; a nested `defaults/<workflow>/...` folder is instead intentionally a packaged workflow whose author-selected pack name is `defaults`, so it is not the same YAML being loaded twice.

File: scripts/generate-bundled-defaults.ts [hunk 2]
```diff
@@ -105,6 +130,23 @@ async function assertNoUntrackedFiles(
...
+  const tracked = new Set(stdout.trim().split('\n').filter(Boolean));
+  const missing = relativePaths.filter(path => !tracked.has(path));
+  if (missing.length > 0) {
+    throw new Error(
+      `Packaged workflows contain untracked files that would be embedded into the binary bundle:\n${missing
+        .map(path => `  ${path}`)
+        .join('\n')}\n\nStage and commit them, or remove them from the packaged workflow folder.`
+    );
+  }
+}
```
[the remark this answers is not in the document]
Comment (Wirasm) [reply scripts/generate-bundled-defaults.ts:200]: Already addressed in `4cdd074`: repository-relative paths are normalized to forward slashes and `git ls-files` runs with `core.quotePath=false`. The Windows Actions job passes with this fix.
